dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/utils/ProgramBuilder.java

```diff
@@ -25,272 +27,307 @@
-            functions.putIfAbsent(tid, new Thread(name, tid, threadEntry));
+    // ----------------------------------------------------------------------------------------------------------------
+    // Misc
+
+    public TypeFactory getTypeFactory() {
+        return types;
+    }
+
+    public ExpressionFactory getExpressionFactory() {
+        return expressions;
+    }
+
+    public void setAssert(AbstractAssert ass) {
+        program.setSpecification(ass);
+    }
+
+    public void setAssertFilter(AbstractAssert ass) {
+        program.setFilterSpecification(ass);
+    }
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // Threads and Functions
+
+    public Thread newThread(String name, int tid) {
+        if(id2FunctionsMap.containsKey(tid)) {
+            throw new MalformedProgramException("Function or thread with id " + tid + " already exists.");
         }
+        final Thread thread = new Thread(name, DEFAULT_THREAD_TYPE, List.of(), tid, EventFactory.newSkip());
+        id2FunctionsMap.put(tid, thread);
+        program.addThread(thread);
+        return thread;
     }
 
-    public Function initFunction(String name, int fid, FunctionType type, List<String> parameterNames) {
-        if(!functions.containsKey(fid)){
-            functions.putIfAbsent(fid, new Function(name, type, parameterNames, fid, null));
-            return functions.get(fid);
+    public Function newFunction(String name, int fid, FunctionType type, List<String> parameterNames) {
+        if(id2FunctionsMap.containsKey(fid)) {
+            throw new MalformedProgramException("Function or thread with id " + fid + " already exists.");
         }
-        return null;
+        final Function func = new Function(name, type, parameterNames, fid, null);
+        id2FunctionsMap.put(fid, func);
+        program.addFunction(func);
+        return func;
     }
 
-    public void initThread(int tid){
-        initThread(String.valueOf(tid), tid);
+    public Thread newThread(int tid) {
+        return newThread(String.valueOf(tid), tid);
     }
 
-    public Event addChild(int fid, Event child) {
-        if(!functions.containsKey(fid)){
-            throw new MalformedProgramException("Function " + fid + " is not initialised");
+    public Thread getOrNewThread(int tid) {
+        if (!id2FunctionsMap.containsKey(tid)) {
+            return newThread(tid);
         }
-        if (child.getThread() != null) {
-            //FIXME: This is a bad error message, but our tests require this for now.
-            final String error = String.format(
-                    "Trying to reinsert event %s from thread %s into thread %s",
-                    child, child.getThread().getId(), fid);
-            throw new MalformedProgramException(error);
+        return (Thread) id2FunctionsMap.get(tid);
+    }
+
+    public boolean functionExists(int fid) {
```

Comment:
not used anywhere. However this class has several instances where this could be used